crates/ruff/src/rules/pylint/rules/redefined_loop_name.rs

```diff
@@ -0,0 +1,194 @@
+    /// an outer loop will reuse the same actual variable, and the value from the last
+    /// iteration will "leak out" into the remainder of the enclosing loop.
+    ///
+    /// While this mistake is easy to spot in small examples, it can be hidden in larger
+    /// blocks of code where the definition and redefinition of the variable may not be
+    /// visible at the same time.
+    ///
+    /// ## Example
+    /// ```python
+    /// for i in range(10):
+    ///     i = 9
+    ///     print(i)  # prints 9 every iteration
+    ///
+    /// for i in range(10):
+    ///     for i in range(10):  # original value overwritten
+    ///         pass
+    ///     print(i)  # also prints 9 every iteration
+    ///
+    /// with path1.open() as f:
+    ///     with path2.open() as f:
+    ///         f = path2.open()
+    ///     print(f.readline())  # prints a line from path2
+    /// ```
+    pub struct RedefinedLoopName {
+        pub name: String,
+    }
+);
+impl Violation for RedefinedLoopName {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let RedefinedLoopName { name } = self;
+        format!("For loop or with statement variable `{name}` overwritten in body")
+    }
+}
+
+struct InnerForWithAssignNamesVisitor<'a> {
+    locator: &'a Locator<'a>,
+    name_ranges: Vec<Range>,
+}
+
+impl<'a, 'b> Visitor<'b> for InnerForWithAssignNamesVisitor<'_>
+where
+    'b: 'a,
+{
+    fn visit_stmt(&mut self, stmt: &'b Stmt) {
+        match &stmt.node {
+            // For and async for.
+            StmtKind::For { target, .. } | StmtKind::AsyncFor { target, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_expr(target, self.locator));
+            }
+            // With.
+            StmtKind::With { items, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_with_items(items, self.locator));
+            }
+            // Assignment, augmented assignment, and annotated assignment.
+            StmtKind::Assign { targets, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_assign_targets(targets, self.locator));
+            }
+            StmtKind::AugAssign { target, .. } | StmtKind::AnnAssign { target, .. } => {
+                self.name_ranges
+                    .extend(name_ranges_from_expr(target, self.locator));
+            }
+            _ => {}
+        }
+        visitor::walk_stmt(self, stmt);
+    }
+}
+
+fn name_ranges_from_expr<'a, U>(target: &'a Expr<U>, locator: &Locator) -> Vec<Range> {
+    find_names(target, locator)
+}
+
+fn name_ranges_from_with_items<'a, U>(items: &'a [Withitem<U>], locator: &Locator) -> Vec<Range> {
+    items
+        .iter()
+        .filter_map(|item| {
+            item.optional_vars
+                .as_ref()
+                .map(|expr| find_names(&**expr, locator))
+        })
+        .flatten()
+        .collect()
```

Comment:
Since these are ultimately passed to `extend`, can this return an iterator rather than a vector?

Reply:
Done - I modified `find_names` to return an iterator too to avoid the vector allocation.